serving/purgatorio/select-and-train-machine-learning-models/deep-learning-theory/deep-learning-theory.md

```diff
@@ -0,0 +1,324 @@
+
+Once you're done with this reading, you can play with some other different explanations in order to deepen your understanding (remember that visualization is awesome for learning, as stated in the [Virgilio's Teaching Strategy](serving\paradiso\virgilio-teaching-strategy\virgilio-teaching-strategy.md) Guide.
+
+Additional resources:
+
+- [Backpropagation Demo](https://google-developers.appspot.com/machine-learning/crash-course/backprop-scroll/)
+- [Backpropagation by 3Blue1Brown](https://www.youtube.com/watch?v=Ilg3gGewQ5U)
+
+_Reinvent the wheel_:
+A good way to understanding backpropagation is to code it from scratch! Try to do it without help, but these tutorials will give you the solution.
+
+- [How to Code a Neural Network with Backpropagation In Python (from scratch)](https://machinelearningmastery.com/implement-backpropagation-algorithm-scratch-python/)
+- [Let's Build a Deep Learning Library](https://www.youtube.com/watch?v=o64FV-ez6Gw)
+
+
+### Transfer Learning:
+
+This is probably one of the most **useful and beautiful** ideas about the world of machines that _learns_ from data.
+
+Transfer learning is a learning strategy in which a model developed for one task is reused as a starting point for a model on a second task. 
+
+It is a popular approach in deep neural network learning, where preformed models are used as a starting point for computer vision and natural language processing activities, given the vast computing resources and time needed to develop neural models on these problems.
+
+Learn the theory behind Transfer Learning in this lecture:
+
+- [Transfer Learning](https://www.youtube.com/watch?v=MIsSuWsZtKE)
+
+Transfer Learning brings with it very useful and powerful features:
+
+- It allows you to train models in a fraction of the expected time.
+
+- Requires much fewer data than "from scratch" training.
+
+- The _diversity_ of the starting dataset (on which the pre-training was done) compared to the dataset on which the fine-tuning is done (the second task) _help the model's generalizing ability_. 
+
+Once you approach new problems, the first thing you should do is to **look for a similar public dataset**, order to find pre-trained models and fine-tune them on your task. Of course, the dataset and the problem need to be _similar_! 
+
+For example, if you want to train a neural network for image classification, you should try to [fine tune a pre-trained model from the Imagenet dataset](https://flyyufelix.github.io/2016/10/03/fine-tuning-in-keras-part1.html). Or, if you are trying some text classification, you should try to [fine tune the BERT model](https://arxiv.org/abs/1905.05583)!
+
+There are a lot of models you can find at the following links:
+
+- [ModelZoo](https://modelzoo.co/)
+- [TensorFlow Hub](https://www.tensorflow.org/hub)
+- [Papers with Code](https://paperswithcode.com/)
+- [State-of-the-art models](https://github.com/onnx/models)
+
+Never forget to try Transfer Learning on your problem!
```

Comment:
Three things to mention here:
- transfer learning may sometimes work when use models that were trained across different domains (i.e. image for text analysis)
- transfer learning works best within the same domain most of the time
- use to make a general model more domain-specific by retraining with new sets of datasets

Reply:
cool. I add these points